pkg/volume/glusterfs/glusterfs.go

```diff
@@ -98,15 +98,15 @@ func (plugin *glusterfsPlugin) GetPluginName() string {
 }
 
 func (plugin *glusterfsPlugin) GetVolumeName(spec *volume.Spec) (string, error) {
-	volumeSource, _, err := getVolumeSource(spec)
+	endpointName, volPath, _, err := getVolumeSource(spec)
 	if err != nil {
 		return "", err
 	}
 
 	return fmt.Sprintf(
 		"%v:%v",
-		volumeSource.EndpointsName,
```

Comment:
what are the uniqueness requirements around what GetVolumeName returns today? why is the namespace not included already?

Reply:
still outstanding

Reply:
this question is still outstanding. what is GetVolumeName used for, and why does it not include the endpoint's namespace today?

Reply:
@liggit apologies for missing this comment.. Here is my thought,  `path` ( ie volume name) is unique for a gluster cluster. That said, only one volume exist in a gluster cluster, so its unique in nature.  But, there is a possibility exist that, second gluster cluster can have the same volume name. Endpoint basically carries gluster nodes IPs , so the uniqueness can be achieved by a combo of endpoint and path.

Reply:
> Endpoint basically carries gluster nodes IPs , so the uniqueness can be achieved by a combo of endpoint and path.

two endpoints with the same name in two different namespaces could reference different IPs. 

example:
* ns1 has endpoint named "gluster" pointing to IP 1.1.1.1, and path `/path/to/volume`
* ns2 has endpoint named "gluster" pointing to IP 2.2.2.2, and path `/path/to/volume`

this method would compute a volume name of "gluster:/path/to/volume" for both, right?

like I said, this isn't new to this PR, but it needs further explanation as to why this is ok. is the namespace information included somehow in the computed path? if so, which namespace is considered, the namespace of the consuming pod, or the namespace of the gluster endpoint?

Reply:
> ns1 has endpoint named "gluster" pointing to IP 1.1.1.1, and path /path/to/volume
    ns2 has endpoint named "gluster" pointing to IP 2.2.2.2, and path /path/to/volume>

Sure.

>this method would compute a volume name of gluster:/path/to/volume for both, right?>

@liggitt, The compute mount will be `1.1.1.1:/path/to/volume` for ns1 and `2.2.2.2:/path/to/volume` for ns2, as we use IPs in endpoint for mount.  I am not 100% sure that,  compute mount path is the only scenario we have to think about though, but survived till now :).

Reply:
> The compute mount will be `1.1.1.1:/path/to/volume` for ns1 and `2.2.2.2:/path/to/volume` for ns2, as we use IPs in endpoint for mount. I am not 100% sure that, compute mount path is the only scenario we have to think about though, but survived till now :).

That's not what this code does. If the endpoint IP gets resolved elsewhere, that's fine, but I'm concerned that GetVolumeName is relying on some other code that is implicitly making use of the pod's namespace to disambiguate the mounts. With the change this PR is making, the pod's namespace is no longer the correct disambiguator in all cases.

Reply:
Sure.  @liggitt  pushing `endpointnamespace` with `endpoint and path` if the persistentvolume spec is non nil in return will do ?

Reply:
`endpointsName:endpointsNamespace:volumePath` (only when pv endpoints namespace is non-nil for backcompat) would be unique and would resolve my concerns about this method. would need an ack from @kubernetes/sig-storage-pr-reviews that that doesn't break assumptions elsewhere (does anything try to parse out endpointsName and path from this volume name elsewhere?)

Reply:
> endpointsName:endpointsNamespace:volumePath (only when pv endpoints namespace is non-nil for backcompat) would be unique and would resolve my concerns about this method>

Sure, will address it in next patch set.

>would need an ack from @kubernetes/sig-storage-pr-reviews that that doesn't break assumptions elsewhere (does anything try to parse out endpointsName and path from this volume name elsewhere?)>

afaik, its not parsed anywhere. the closest code is  'reconstructvolume' , but reconstuction is not supported (https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/glusterfs/glusterfs.go#L199) for glusterfs. `GetVolumeName()` is mainly used for `attach/detach` volumes and glusterfs is not into this category.  @jsafrane @rootfs can you confirm my thoughts  or share details?

Reply:
@liggitt added endpointsnamespace as mentioned. PTAL :) .. Thanks !!!

Reply:
cc @msau42 on modification to GetVolumeName()

Reply:
GetVolumeName() is used in a few places, all of which are not applicable to gluster:
* AD controller
* kubelet volume manager actual/desired sw for attachable volumes
* volume reconstruction

Can we just return error from GetVolumeName() since it shouldn't even be called?

Reply:
Since this function should not come into play for gluster, backwards compatibility should not be a concern.  I am fine with cleaning this up in a followup.